File: src/agent_nexus/platform/agency/cli_backend/registry.py
"""CLIBackendRegistry — backend discovery, registration, and health check."""

from __future__ import annotations

import logging
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from agent_nexus.platform.agency.cli_backend.base import GenericCLIBackend

logger = logging.getLogger(__name__)


class CLIBackendRegistry:
    """Registry of available CLI backends with health check support."""

    def __init__(self) -> None:
        self._backends: dict[str, GenericCLIBackend] = {}

    def register(self, name: str, backend: GenericCLIBackend) -> None:
        self._backends[name] = backend
        logger.debug("Registered CLI backend: %s -> %s", name, backend.name)

    def get(self, name: str) -> GenericCLIBackend:
        if name not in self._backends:
            raise KeyError(
                f"CLI backend '{name}' not registered. Available: {list(self._backends.keys())}"
            )
        return self._backends[name]

    def available_backends(self) -> list[GenericCLIBackend]:
        return [b for b in self._backends.values() if b.is_available()]

    def all_backends(self) -> dict[str, GenericCLIBackend]:
        return dict(self._backends)

    def refresh_all(self) -> None:
        for backend in self._backends.values():
            backend.refresh_availability()

    def __len__(self) -> int:
        return len(self._backends)

File: crates/ap-cli-backend/src/registry.rs
//! CLIBackendRegistry — backend discovery and health check.

use crate::backend::GenericCLIBackend;
use std::collections::HashMap;
use std::sync::Arc;

#[derive(Clone)]
pub struct CLIBackendRegistry {
    backends: HashMap<String, Arc<GenericCLIBackend>>,
}

impl Default for CLIBackendRegistry {
    fn default() -> Self {
        Self::new()
    }
}

impl CLIBackendRegistry {
    pub fn new() -> Self {
        Self { backends: HashMap::new() }
    }

    pub fn register(&mut self, name: String, backend: GenericCLIBackend) {
        self.backends.insert(name, Arc::new(backend));
    }

    pub fn get(&self, name: &str) -> Result<Arc<GenericCLIBackend>, String> {
        self.backends.get(name)
            .cloned()
            .ok_or_else(|| format!("CLI backend '{}' not registered", name))
    }

    pub fn available_backends(&self) -> Vec<Arc<GenericCLIBackend>> {
        self.backends.values()
            .filter(|b| b.is_available())
            .cloned()
            .collect()
    }

    pub fn len(&self) -> usize {
        self.backends.len()
    }
}
